zcash_transparent/src/bundle.rs

```diff
@@ -258,6 +267,72 @@ impl TxIn<Authorized> {
     }
 }
 
+impl TxIn<builder::Coinbase> {
+    /// Creates an input for a coinbase transaction. Does not support creating inputs for the
+    /// genesis block.
+    pub fn coinbase(height: BlockHeight, miner_data: &MinerData) -> Result<Self, coinbase::Error> {
+        let mut script_sig = match i64::from(height) {
+            0 => Err(coinbase::Error::GenesisInputNotSupported)?,
+            h => push_num(h).to_bytes(),
+        };
+
+        // # Consensus
+        //
+        // > A coinbase transaction for a block at block height greater than 0 MUST have a script
+        // > that, as its first item, encodes the block height `height` as follows. For `height` in
+        // > the range {1 .. 16}, the encoding is a single byte of value `0x50` + `height`.
+        // > Otherwise, let `heightBytes` be the signed little-endian representation of `height`,
+        // > using the minimum nonzero number of bytes such that the most significant byte is <
+        // > `0x80`. The length of `heightBytes` MUST be in the range {1 .. 5}. Then the encoding is
+        // > the length of `heightBytes` encoded as one byte, followed by `heightBytes` itself. This
+        // > matches the encoding used by Bitcoin in the implementation of [BIP-34] (but the
+        // > description here is to be considered normative).
+        //
+        // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+        //
+        // ## Note
+        //
+        // The height is encoded in the `push_num` fn above.
+        //
+        // [BIP-34]: <https://github.com/bitcoin/bips/blob/master/bip-0034.mediawiki>
+        if script_sig.is_empty() {
+            Err(coinbase::Error::EmptyHeight)?;
+        } else if script_sig.len() > coinbase::MAX_COINBASE_HEIGHT_LEN {
+            Err(coinbase::Error::OversizedHeight)?;
+        }
+
+        // We can embed `miner_data` directly into `script_sig` since it is already encoded as a
+        // push value.
+        script_sig.extend(miner_data.as_ref());
+
+        // # Consensus
+        //
+        // > A coinbase transaction script MUST have length in {2 .. 100} bytes.
+        //
+        // <https://zips.z.cash/protocol/protocol.pdf#txnconsensus>
+        //
+        // ## Note
+        //
+        // These bounds are met implicitly, but we check explicitly to document compliance with the
+        // consensus rule.
+        if script_sig.len() < coinbase::MIN_COINBASE_SCRIPT_LEN {
+            Err(coinbase::Error::UndersizedScript)?;
+        } else if script_sig.len() > coinbase::MAX_COINBASE_SCRIPT_LEN {
+            Err(coinbase::Error::OversizedScript)?;
+        }
```

Comment:
To avoid calculating the length more than once from a `script::Sig`:
```suggestion
        match script_sig.byte_len() {
            0..coinbase::MIN_COINBASE_SCRIPT_LEN => Err(coinbase::Error::UndersizedScript)?,
            coinbase::MIN_COINBASE_SCRIPT_LEN..=coinbase::MAX_COINBASE_SCRIPT_LEN => (),
            (coinbase::MAX_COINBASE_SCRIPT_LEN + 1).. =>Err(coinbase::Error::OversizedScript)?,
        }
```

Reply:
#2078 also uses only a single call for the length.